### PR TITLE
MINOR: refactor BuiltInPartitioner to remove mockRandom from production code

### PR DESCRIPTION
**Pull request by gongxuanzhang**
The original author added a note in `BuiltInPartitioner`  
> Visible and used for testing only

I think this will break the scope.
After compilation, the test module and the source code are in the same package.
So we don't need `public`, just need `protected`

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java [hunk 1]
```diff
@@ -45,7 +45,7 @@ public class BuiltInPartitioner {
     private final AtomicReference<StickyPartitionInfo> stickyPartitionInfo = new AtomicReference<>();
 
     // Visible and used for testing only.
-    static volatile public Supplier<Integer> mockRandom = null;
+    protected static volatile Supplier<Integer> mockRandom = null;
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java]: Maybe a better way is to add a package-private method `random()`, and then in the `BuiltInPartitionerTest`  we create a `BuiltInPartitioner` with override `random()`. With that change we can remove "mock" from `BuiltInPartitioner`.

Comment (gongxuanzhang) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java]: `mockRandom` is a static field, which means its behavior can be modified anywhere in the test. There are many instances in the context, not just within the test packages. Therefore, if we add a function and let the test module extend this class, it will cause the already created objects to not work as expected, requiring significant code changes.

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java]: Agree that could introduce some changes, but we can minimize the changes by opening only required methods. For example:
**BuiltInPartitioner**
```java
    int random() {
        return Utils.toPositive(ThreadLocalRandom.current().nextInt());
    }

    private int nextPartition(Cluster cluster) {
        int random = random();
        ...
    }
```
**RecordAccumulator**
```java
    private Deque<ProducerBatch> getOrCreateDeque(TopicPartition tp) {
        TopicInfo topicInfo = topicInfoMap.computeIfAbsent(tp.topic(), k -> new TopicInfo(createBuiltInPartitioner(logContext, k, batchSize)));
        return topicInfo.batches.computeIfAbsent(tp.partition(), k -> new ArrayDeque<>());
    }
    BuiltInPartitioner createBuiltInPartitioner(LogContext logContext, String topic, int stickyBatchSize) {
        return new BuiltInPartitioner(logContext, topic, stickyBatchSize);
    }

        public TopicInfo(BuiltInPartitioner builtInPartitioner) {
            this.builtInPartitioner = builtInPartitioner;
        }
```

With those override-able methods, both `RecordAccumulatorTest` and `BuiltInPartitionerTest` can create `BuiltInPartitioner` with custom `random` method. WDYT?

Comment (gongxuanzhang) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java]: I modified this PR.
BTW,I moved the hard-coded parts to a private class within the test module in order to pass the original tests

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 2]
```diff
@@ -1763,4 +1747,65 @@ private RecordAccumulator createTestRecordAccumulator(
             txnManager,
             new BufferPool(totalSize, batchSize, metrics, time, metricGrpName));
     }
+
+    private class SequentialRecordAccumulator extends RecordAccumulator {
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: Could we move `AtomicInteger mockRandom` to class-level and then return `RecordAccumulator` with override  `createBuiltInPartitioner`? For example:
**RecordAccumulatorTest**
```java
    private AtomicInteger random = null;

    @AfterEach
    public void teardown() {
        this.metrics.close();
    }

    private RecordAccumulator createTestRecordAccumulator(
        TransactionManager txnManager,
        int deliveryTimeoutMs,
        int batchSize,
        long totalSize,
        Compression compression,
        int lingerMs
    ) {
        long retryBackoffMs = 100L;
        long retryBackoffMaxMs = 1000L;
        String metricGrpName = "producer-metrics";

        return new RecordAccumulator(
            logContext,
            batchSize,
            compression,
            lingerMs,
            retryBackoffMs,
            retryBackoffMaxMs,
            deliveryTimeoutMs,
            metrics,
            metricGrpName,
            time,
            new ApiVersions(),
            txnManager,
            new BufferPool(totalSize, batchSize, metrics, time, metricGrpName)) {
            @Override
            BuiltInPartitioner createBuiltInPartitioner(LogContext logContext, String topic, int stickyBatchSize) {
                return new BuiltInPartitioner(logContext, topic, stickyBatchSize) {
                    @Override
                    int random() {
                        return random == null ? super.random() : random.getAndIncrement();
                    }
                };
            }
        };
    }
```

With that change, `testUniformBuiltInPartitioner` can define the `random` in the begining
```java
random = new AtomicInteger();
```

`testAdaptiveBuiltInPartitioner` can create `RecordAccumulator` with override `createBuiltInPartitioner` directly.
```java
        RecordAccumulator accum = new RecordAccumulator(logContext, batchSize, Compression.NONE, 0, 0L, 0L,
            3200, config, metrics, "producer-metrics", time, new ApiVersions(), null,
            new BufferPool(totalSize, batchSize, metrics, time, "producer-internal-metrics")) {
            @Override
            BuiltInPartitioner createBuiltInPartitioner(LogContext logContext, String topic, int stickyBatchSize) {
                return new BuiltInPartitioner(logContext, topic, stickyBatchSize) {
                    @Override
                    int random() {
                        return random == null ? super.random() : random.getAndIncrement();
                    }
                };
            }
        };
```

Comment (gongxuanzhang) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: This is an excellent suggestion!, perfectly reproducing the "global modification" in the previous code.
I changed PR, plz review it @chia7712

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java [hunk 3]
```diff
@@ -113,6 +110,10 @@ private int nextPartition(Cluster cluster) {
         return partition;
     }
 
+    protected int randomPartition() {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java]: Could you change it to package-private and add `visible for testing`?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java [hunk 4]
```diff
@@ -1033,9 +1033,17 @@ public Deque<ProducerBatch> getDeque(TopicPartition tp) {
...
-        TopicInfo topicInfo = topicInfoMap.computeIfAbsent(tp.topic(), k -> new TopicInfo(logContext, k, batchSize));
+        TopicInfo topicInfo = topicInfoMap.computeIfAbsent(tp.topic(),
+                k -> new TopicInfo(createBuiltInPartitioner(logContext, k, batchSize)));
         return topicInfo.batches.computeIfAbsent(tp.partition(), k -> new ArrayDeque<>());
     }
+    
+    /**
+     * Subclass can custom {@link BuiltInPartitioner}
+     **/
+    protected BuiltInPartitioner createBuiltInPartitioner(LogContext logContext, String topic, int stickyBatchSize) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java]: Could you change it to package-private and add visible for testing?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 5]
```diff
@@ -103,13 +103,13 @@ public class RecordAccumulatorTest {
...
     private MetadataSnapshot metadataCache = new MetadataSnapshot(null,
-        nodes,
-        partMetadatas,
-        Collections.emptySet(),
-        Collections.emptySet(),
-        Collections.emptySet(),
-        null,
-        Collections.emptyMap());
+            nodes,
+            partMetadatas,
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: please revert unrelated changes

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 6]
```diff
@@ -146,17 +148,17 @@ public void testDrainBatches() throws Exception {
         PartitionInfo part3 = MetadataResponse.toPartitionInfo(partMetadata3, nodes);
         PartitionInfo part4 = MetadataResponse.toPartitionInfo(partMetadata4, nodes);
         Cluster cluster = new Cluster(null, Arrays.asList(node1, node2), Arrays.asList(part1, part2, part3, part4),
-            Collections.emptySet(), Collections.emptySet());
+                Collections.emptySet(), Collections.emptySet());
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 7]
```diff
@@ -121,12 +121,14 @@ public class RecordAccumulatorTest {
...
     private final long maxBlockTimeMs = 1000;
     private final LogContext logContext = new LogContext();
+    private AtomicInteger mockRandom = null;
 
     @BeforeEach void setup() {}
 
     @AfterEach
     public void teardown() {
         this.metrics.close();
+        mockRandom = null;
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: this is redundant since the variable is case-level

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 8]
```diff
@@ -146,17 +148,17 @@ public void testDrainBatches() throws Exception {
...
-            Collections.emptySet(),
-            Collections.emptySet(),
-            Collections.emptySet(),
-            null,
-            Collections.emptyMap(),
-            cluster);
+                nodes,
+                partMetadatas,
+                Collections.emptySet(),
+                Collections.emptySet(),
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 9]
```diff
@@ -1737,30 +1735,48 @@ private RecordAccumulator createTestRecordAccumulator(int deliveryTimeoutMs, int
...
+                new BufferPool(totalSize, batchSize, metrics, time, metricGrpName)) {
+            @Override
+            protected BuiltInPartitioner createBuiltInPartitioner(LogContext logContext, String topic,
+                                                                  int stickyBatchSize) {
+                return new MockRandomBuiltInPartitioner(logContext, topic, stickyBatchSize);
+            }
+        };
+    }
+
+    private class MockRandomBuiltInPartitioner extends BuiltInPartitioner {
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: it can be static class, and it would be great to unify the naming -> `SequentialPartitioner`

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/BuiltInPartitionerTest.java [hunk 10]
```diff
@@ -203,4 +189,19 @@ void testStickyBatchSizeMoreThatZero() {
         assertThrows(IllegalArgumentException.class, () -> new BuiltInPartitioner(logContext, TOPIC_A, 0));
         assertDoesNotThrow(() -> new BuiltInPartitioner(logContext, TOPIC_A, 1));
     }
+    
+    
+    static class SequentialPartitioner extends BuiltInPartitioner {
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/BuiltInPartitionerTest.java]: could you please make it be private ?

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang the idea is good to me. Please notice that unrelated changes :)

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java (hunk 9 above)
Comment (gongxuanzhang) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: It needs to access the instance properties of the test class so can't be static class

Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: Got it. thanks!

**Comment by gongxuanzhang** [comment]
I updated it ,plz review it.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java [hunk 11]
```diff
@@ -1033,10 +1033,18 @@ public Deque<ProducerBatch> getDeque(TopicPartition tp) {
...
      */
     private Deque<ProducerBatch> getOrCreateDeque(TopicPartition tp) {
-        TopicInfo topicInfo = topicInfoMap.computeIfAbsent(tp.topic(), k -> new TopicInfo(logContext, k, batchSize));
+        TopicInfo topicInfo = topicInfoMap.computeIfAbsent(tp.topic(),
+                k -> new TopicInfo(createBuiltInPartitioner(logContext, k, batchSize)));
         return topicInfo.batches.computeIfAbsent(tp.partition(), k -> new ArrayDeque<>());
     }
 
+    /**
+     * Subclass can custom {@link BuiltInPartitioner}
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java]: this comment is a bit redundant, since it just describe what the code. It would be better to explain why we need this method rather than inline it. Or we can remove the docs

Comment (gongxuanzhang) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java]: You are right,it seem to don’t need explaining. I removed docs
